### DISCUSSION TITLE
File-based scoping

### DISCUSSION THREAD

**Discussion by drhagen** [eclipse-langium/langium#1451] [Q&A] [answered]
Are there any examples of scoping names to the file? Neither of [scoping guides](https://langium.org/guides/scoping/) does this. I don't want to do gimme teh codez, but this is a common enough language feature that I am guessing someone out there has a DSL with file-based scoping already.

**Comment by cdietrich** [comment]
What do you mean by that? No references to other files?

**Reply by drhagen** [reply]
Sure, for simplicity, let's assume none at all. A variable `foo` can reference something defined in the current file, but does not see `foo` declared in a different file.

**Reply by cdietrich** [reply]
https://github.com/eclipse-langium/langium/blob/f3321b6e65d8120db85ebec1c3d2b3583ea7568d/packages/langium/src/references/scope-provider.ts#L101
override and return empty

**Reply by drhagen** [reply, marked as answer]
Overriding `getGlobalScope` and returning an empty scope works. Here's the minimal code for modifying the Hello World project to enforce file-scoping.

Add file `hello-world-scope-provider.ts` with contents:

```typescript
import type { Scope } from 'langium';
import { EMPTY_SCOPE } from 'langium';
import { DefaultScopeProvider } from 'langium';

/**
 * Scope provider that restricts scope to a single file
 */
export class HelloWorldScopeProvider extends DefaultScopeProvider {
    protected override getGlobalScope(referenceType: string): Scope {
        return EMPTY_SCOPE;
    }
}
```

Like in the other examples modify `HelloWorldModule` in `hello-world-module.ts` to inject this scope provider:

```typescript
// ...
export const HelloWorldModule: Module<HelloWorldServices, PartialLangiumServices & HelloWorldAddedServices> = {
    references: {
        ScopeProvider: (services) => new HelloWorldScopeProvider(services)
    },

// ...
};
// ...
```
